PLT-7342: Implement layout and landing pages as per improved layout

EehMauro opened the pull request:
https://input-output.atlassian.net/browse/PLT-7342

BE Branch : `feat/change-profile-fields`, [PR #94](https://github.com/input-output-hk/dapps-certification/pull/94) --> [PLT-6187](https://input-output.atlassian.net/browse/PLT-6187)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
### Summary by CodeRabbit

- New Feature: Introduced a new landing page with registration and subscription sections.
- New Feature: Added a navigation bar component for better user navigation.
- New Feature: Implemented wallet connection functionality for login/sign-up.
- New Feature: Created a Redux slice for handling registration-related actions.
- Refactor: Restructured the routing logic in the `App` component.
- Refactor: Updated the `auth.slice` to include additional properties and async thunks.
- Refactor: Removed unused components and imports, simplifying the codebase.
- Chore: Added Tailwind CSS for styling various components in the application.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

@coderabbitai: ignore

amnambiar committed:
PLT-7342 revamped layout

Commit:
Replaced @mui/styles with tss-react to fix React 18 compatibility

Commit:
Added tailwind css, removed reset css and using mui css baseline

Commit:
Added landing page

RSoulatIOHK commented on src/components/NavBar/NavBar.tsx:
```diff
@@ -0,0 +1,54 @@
+        }}
+        className={bgColors.classes["bg-purple-dark"]}
+      >
+        Certifyr.io
```
```suggestion
        Testing Tool
```

We got a red light on the naming so for now, we'll just use a generic name

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,181 @@
+    <Box className="flex flex-col justify-center m-0 min-w-[50vw] bg-white">
+      <Box className="text-center">
+        <Typography variant="h3" className="font-bold mb-4 tracking-[.2em]">
+          Certifyr.io
```
```suggestion
          Testing Tool
```
We got a red light on the naming so for now, we'll just use a generic name

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,181 @@
+            Developer<br/>Premium Tier
+          </Typography>
+          <Typography variant="h2" className="font-normal mb-24">
+            $999
```
Shouldn't the price be pulled from the subscription table?

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,181 @@
+        Please choose the Package you wish to subscribe to.
+      </Typography>
+      <Box className="flex flex-row px-2">
+        <Box className={props.packageType === 'developer' ? 'package-type-active' : 'package-type'}>
+          <Typography variant="h5" className="font-bold my-8">
+            Developer<br/>Premium Tier
+          </Typography>
+          <Typography variant="h2" className="font-normal mb-24">
+            $999
+          </Typography>
+          <Typography variant="body1" className="font-bold">
+            Limited testing features
+          </Typography>
+          <Typography variant="body1" className="font-normal mb-12">
+            Limited number of checks<br />Limited number of properties<br />Low priority queue
+          </Typography>
+          <Button
+            variant="contained"
+            size="large"
+            className="py-3 px-14"
+            color={!props.packageType || props.packageType === 'developer' ? 'info' : 'primary'}
+            onClick={() => props.onSelect('developer')}
+          >
+            {props.packageType === 'developer' ? 'Selected' : 'Select'}
+          </Button>
+        </Box>
```
Could we disable this package from being selected?

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,181 @@
+            Auditor<br/>Premium Tier
+          </Typography>
+          <Typography variant="h2" className="font-normal mb-24">
+            $9,999
```
Shouldn't the price be pulled from the subscription table?

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,181 @@
+            $9,999
+          </Typography>
+          <Typography variant="body1" className="font-bold">
+            Unlimited testing features
```
```suggestion
            Unlimited testing (1 run at a time)
```

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,181 @@
+            Unlimited testing features
+          </Typography>
+          <Typography variant="body1" className="font-normal mb-12">
+            Unlimited number of checks<br />Unlimited number of properties<br />High priority queue
```
```suggestion
            Unlimited number of checks<br />Unlimited number of properties<br />High priority queue <br /> Level 0 and 2 certificate builder
```

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,181 @@
+          className="mt-8 py-3 px-14 normal-case"
+          onClick={props.onSubmit}
+        >
+          Pay (₳3,181.08)
```
Same, the price should be computed from the price of subscription in usd converted to the price in Ada with the rate ada/usd

Commit:
Added fetch tiers, mocked fetch price and registation form validations

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,201 @@
+        <Typography variant="h3" className="font-bold mb-4 tracking-[.2em]">
+          Testing Tool
+        </Typography>
+        <Typography variant="h6" className="font-normal mb-40 text-[#A6A7AD]">
+          Connect your in-browser wallet to login/sign-up
+        </Typography>
+        <Button variant="contained" size="large" className="py-3 px-4 normal-case font-medium" onClick={props.onSubmit}>
+          Connect your wallet
+        </Button>
```
There are 2 buttons for connect wallet. Is it while we finish everything to have the two routes?

EehMauro replied:
Yes, I'm finishing the integration right now, it'll be fixed in the next commit

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,201 @@
+        Testing Tool
+      </Typography>
+      <Typography variant="subtitle1" className="font-normal mb-16 text-[#A6A7AD]">
+        Please choose the Package you wish to subscribe to.
```
```suggestion
        Please choose a subscription
```

I don't like the final period and I didn't like the wording. If you think this is worse, disregard it, I don't have a strong feeling about this ^^

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,201 @@
+      Auditor profile
+    </Typography>
+    <Typography variant="subtitle1" className="font-normal mb-16 text-[#A6A7AD]">
+      Please complete the information to create your account.
```
```suggestion
      Please complete your user profile information
```

Same as previous comment.

RSoulatIOHK commented on src/pages/landing/Landing.tsx:
```diff
@@ -0,0 +1,201 @@
+          Pay (₳{props.price.toFixed(2)})
+        </Button>
+        <Typography variant="body1" className="font-normal text-sm mt-6 text-red-500">
+          Price will refresh in {props.count} seconds
```
I'm nit picking, don't do it if it takes more than 5 seconds to fix.

Can we have a function like this one:
```
function s_of_int(n: number): string {
  return n !== 1 ? "s" : "";
}
```
so that it displays "Price will refresh in 1 second" instead of "1 seconds"?

EehMauro replied:
Sure!

RSoulatIOHK commented on public/images/landing-background.svg:
The image I used might 
not be free to use. I have a new one: 
![neons-degrades-abstraits](https://github.com/input-output-hk/dapps-certification-web/assets/117812549/f5a5d958-e659-4428-9315-efc1e74a14f0)

I also have it as .svg but it's 17MB and GitHub doesn't like it (and maybe we don't as well as it's a bit too big ^^)

Commit:
Updated wallet integration

Commit:
Updated landing integration

EehMauro commented:
A little summary of the latest changes:
```
New changes:
- Fixed the app routes
  - Now when you don't have an active subscription (l2-upload-report & l1-run) you'll be redirected to / (landing)
  - If you have an active subscription and go to / you'll be redirected to /home
- The tiers info now is fetched when you connect your wallet, and the developer tier is disabled
- The subscription form now has the patterns validations with react-hook-form
- Improved the tailwindcss theming, adding the colors to the config file, and adding more styles instead of inline classes
- The landing page components are now modularized, so they're in separated files and they're more easy to read and understand their responsibility
- The ADA USD price now is fetched from the API and used for the subscription cost calculation
- More custom components now are replaced with the MaterialUI ones

What's missing:
- I've to check the PUT /profile/current, looks like I'm not sending the fields right and I'm receiving an 400 error
- I've to add the listeners to detect when the wallet is changed, so I can clear the auth token an redirect the user to connect again
- More testing. After I finish these two things I've to test everything a little more

Nice to have:
- I can improve the wallet connection component, there are some HTTPS requests done inside of the component and they can be moved to a redux action
- I should update the NavBar component, right now it's not looking right because I've been getting rid of styles and replacing everything with MaterialUI components
```

amnambiar reviewed (changes requested):
@EehMauro 

Few corrections and suggestions for now - 

1. Need to incorporate the **Pay** for subscription from existing Payment.tsx
  - Note the Subscription will not be right away active after payment. Any payment takes time to be reflected. So we have to keep polling for the status to make sure it is active.
  - Note that all payments in our tool have to made checking/verifying if the user's profile already has some balance in it. But in this case, since it is in the creation stage, we can ignore those respective snippets within Payment.tsx where it check for available balance, and only then triggers payment.
  - summarising workflow with methods in Payment.tsx --- on Click of Pay button, `postSubscription()` -> `fetchCurrentSubscription()` -> `triggerPayment()` -> `triggerTransactionFromWallet`([<min-fee/fee>](https://github.com/input-output-hk/dapps-certification-web/blob/PLT-7342/src/pages/subscription/payment/Payment.tsx#L76-L81)) 

2. On `active` of subscription, we still might have to show a confirmation of the payment with the transaction-link. Maybe, we need to show this as he comes into the tool - when in Home from a new profile creation, show a modal with _View your performed payment transaction for Subscription to our services, here `https://preprod.cardanoscan.io/transaction/<transactionId>`_

3. Instead of `/home` route, can we handle `sessionFetched`/`hasAnActiveSubscription` within `LandingPage` ;  and chose to render `<Home />` component when active

4. Correct the request payload structure for [PUT /profile/current](https://excuse.ro/swagger-ui/#/default/put_profile_current) - to fix the 400 Error
![image](https://github.com/input-output-hk/dapps-certification-web/assets/11058223/12f0b3c2-edc2-4343-ba50-9789ddef5678)
  -  Optional fields (dapp, twitter, linkedin) should not be included in the request payload, if they are not filled in the form/empty

Commit:
Updated redux slices

Commit:
Finished integrations

Commit:
Added wallet changes integration

Commit:
Fixed integration

Commit:
Updated navigation

Commit:
Updated pages and component from master

EehMauro commented:
@coderabbitai ignore

EehMauro commented:
What's new:
- The connect wallet now is fully integrated with the landing page
- The subscription page now is listening when the wallet or network changes and logout the user if this happens
- The payment feature now is connected to the landing page
- The routes are working again, and the user session is detected
- The navbar is working again

What's missing:
- Some pages are failing now because of the changes on the redux slices
- Some pages are not rendering well because of the changes on the navbar and new app styles

Commit:
Added header

Commit:
Merge branch 'master' into PLT-7342

Commit:
Merged PLT-7455 changes (Switch to stake-address & Yoroi support)

amnambiar commented:
@RSoulatIOHK ,

We planned to fix the respective failing pages, on menu navigation, while working on those pages. Hence, meanwhile we can consider this PR to be Connect,Subscription/Profile,Menu-Header UI, logout(on click/wallet change/network change) and review tool-workflow

**Open Bug:** 
_Planning to work on it, as a bug (not in this PR for now so we can move on to other tasks)_ 

> If for some reason, subscription stayed in "pending" state and the http-request/API failed - this case is not handled now.

Need to handle this.
Suggested Plan - On any such API errors/polling subscription state for 30 seconds, show an error stating "Something wrong occurred. Please revisit our tool to confirm the subscription" ; and take to logout state. When he login, first check for /active-features (should be [] if no subscription or even if in "pending" state); if [], get subscription state; if still "pending", don't show form or ask to pay, instead something that says a "Subscription you purchased is still being processed. Please wait here until it is active, or **contact our team**" [because i guess there is no other way he can resolve it]

amnambiar commented:
@EehMauro , 

One suggestion - to stop polling for /ada-usd-price and refetching price in Pay, after clicking on Pay and while transaction is in progress.

amnambiar commented:
@EehMauro , 

If profile has balance in it (due to a failed subscription attempt earlier), but not enough to buy a subscription ; now the logic in register.slice doesn't allow the user to pay and subscribe. It throws an error "Not enough balance'. We need to recheck the logic.

Be the payment being made from the available profile balance or the wallet transaction, we still need to **poll** for the latest status of Subscription until it is active. See **L86** and **L100** in [Payment.tsx](https://github.com/input-output-hk/dapps-certification-web/blob/8b297162aec2108134f101e1db7e334b71a28918/src/pages/subscription/payment/Payment.tsx)

amnambiar committed:
PLT-7342 Removed Certification and Added Auditor Report Upload with icon

amnambiar commented:
@EehMauro ,

Few others I came across - 

1. In address on top-left of header, we should be showing the changeAddress, rather than the stakeAddress.
2. Can't see the warning banner asking user to ensure himself if he is connected to Preprod ([see here](https://github.com/input-output-hk/dapps-certification-web/blob/8b297162aec2108134f101e1db7e334b71a28918/src/app/App.tsx#L37))

Commit:
Added node-sass

Commit:
Resolved PRs comments

amnambiar commented on src/store/slices/register.slice.ts:
```diff
@@ -37,52 +37,87 @@ const initialState: RegisterState = {
-    if (paymentRes?.error?.message) throw paymentRes?.error;
-    const transactionId = paymentRes.payload;
+      const { isBalanceEnough, fee } = await calculateFee(subscriptionPrice, balance);
+      if (!isBalanceEnough) throw { message: 'Not enough balance' };
```
In this condition, when is the user prompted to actually perform the transaction; in a case where he some balance in his account?
@EehMauro 

NB: I will try to add a fix for this

amnambiar replied:
I renamed the variable isBalanceEnough to lessBalance (otherwise the inversion of values should be stored in it to match the variable meaning). 

Also, when the balance in account is enough to match the tier-price, we should not throw error, and simply allow the tool to continue the payment of subscription from the balance.

amnambiar commented on src/pages/session/components/AppBar.tsx:
```diff
@@ -14,11 +15,9 @@ const AppBar = () => {
   const getProfileAddress = () => {
-    if (!profile || !profile.address) return '...';
-    return `${profile.address.slice(0,4)}...${profile.address.slice(-4)}`;
+    const address = localStorage.getItem(LocalStorageKeys.address);
```
@EehMauro , 
We need to make use of useLocalStorage hook for modularisation setting and getting values from localStorage.

amnambiar commented on src/pages/session/components/AppLayout.tsx:
```diff
@@ -12,10 +11,10 @@ import NavBar from './NavBar';
 
 const Banner = () => {
-  const { networkId } = useAppSelector((state) => state.auth);
+  const networkId = localStorage.getItem(LocalStorageKeys.networkId);
```
same here; to use useLocalStorage hook

amnambiar commented on src/store/slices/auth.slice.ts:
```diff
@@ -112,7 +112,10 @@ export const connectWallet = createAsyncThunk('connectWallet', async (payload: {
+
     localStorage.setItem(LocalStorageKeys.authToken, loginRes.data);
+    localStorage.setItem(LocalStorageKeys.address, changeAddressBech32);
+    localStorage.setItem(LocalStorageKeys.networkId, networkId);
```
same here; to use useLocalStorage hook

Additionally, while replacing with the hook learned that - `Reducers have to be pure functions. That means no side effects like reading data from anywhere except from store and action and no writing data anywhere, too. You are not allowed to use localStorage from a reducer.`. Hence rewriting this a bit.

amnambiar replied:
Upon trying to resolve this, for some reason the `authToken` is not being set properly into the localStorage/state/request-headers. Hence leaving these bit as it.

amnambiar committed:
PLT-7342 To handle the case where subscriptions array is empty (case when pending subscription is deleted through the API)

amnambiar commented on src/store/slices/register.slice.ts:
```diff
@@ -0,0 +1,164 @@
+  return BigNum.from_str(res.data.toString());
+}
+
+const calculateFee = async (subscriptionPrice: BigNum, balance: BigNum) => {
+  let isBalanceEnough = false;
+  if (balance.less_than(subscriptionPrice)) {
+    isBalanceEnough = true;
+  } else {
+    const difference = balance.checked_sub(subscriptionPrice);
+    isBalanceEnough = difference.less_than(BigNum.from_str('0'));
+  }
+
+  const oneAdaInLovelaces = 1000000;
+  let fee = BigNum.from_str(oneAdaInLovelaces.toString()); // 1 ADA in lovelaces - min req for cardano wallet txn
+  if (fee.less_than(subscriptionPrice)) fee = subscriptionPrice;
+
+  return { isBalanceEnough, fee };
+}
```
renamed isBalanceEnough to lessBalance, the logic is correct

amnambiar committed:
PLT-7342 Renamed balance variable and other text

amnambiar commented on src/store/slices/register.slice.ts:
```diff
@@ -0,0 +1,164 @@
+          await new Promise(resolve => setTimeout(resolve, 1000));
+        }
+      } catch (error: any) {}
+    }
```
we have to add this, but sync with BE as well first. Can cover as a separate bug fix

amnambiar commented on src/store/slices/register.slice.ts:
```diff
@@ -0,0 +1,168 @@
+  return BigNum.from_str(res.data.toString());
+}
+
+const calculateFee = async (subscriptionPrice: BigNum, balance: BigNum) => {
+  let lessBalance = false;
+  if (balance.less_than(subscriptionPrice)) {
+    lessBalance = true;
+  } else {
+    const difference = balance.checked_sub(subscriptionPrice);
+    lessBalance = difference.less_than(BigNum.from_str('0'));
+  }
+
+  const oneAdaInLovelaces = 1000000;
+  let fee = BigNum.from_str(oneAdaInLovelaces.toString()); // 1 ADA in lovelaces - min req for cardano wallet txn
+  if (fee.less_than(subscriptionPrice)) fee = subscriptionPrice;
+
+  return { lessBalance, fee };
+}
```
leaving it so for readability and understanding the idea behind the conditions

Commit:
Resolved comments

Commit:
Fixed hook call

Commit:
Fixed session local storage access

Commit:
Fixed scrolling and report upload page

EehMauro commented:
Changes:
- Added missing styles on tailwind (rotate animation, disabled, link)
- Fixed tier selected button styles
- Fixed app network banner 
- Fixed app scrolling
- Fixed Auditor page rendering error

RSoulatIOHK commented on src/pages/landing/components/RegisterSection.tsx:
```diff
@@ -0,0 +1,168 @@
+              className="mt-8 py-3 px-14 normal-case bg-main"
+              disabled={total <= 0 || processing}
+            >
+              Pay (₳{total.toFixed(2)})
```
Wrong price in ada displayed. I see 0.50A (roughly 2Ada, I guess). The conversion should be in the other order.
Also the price should be fetched from the tier table.

EehMauro replied:
Ok, from the `/tiers` the price I get is `"usdPrice": 2` and from the `/ada-usd-price` I get `0.246953`
That's why I got `2 * 0.246953 = 0.493906 A`, but I'm seeing that the price in the subscription is `8098707`
So I don't know how to get the value for the payment button without having the subscription
What calculation should I do?

RSoulatIOHK replied:
The price you get from the tier is in USD and the ada-usd-price is the price of 1 Ada in USD. So you have to who 2/Ada-use-price, which is 2 / 0.246953 = 8.098707A. Which should be displayed, rounded up after 2 decimals, so here display 8.10A.

The price requested by the wallet is in Lovelaces and 1Ada = 1,000,000 Lovelaces so in this case 8,098,707 Lovelaces

RSoulatIOHK commented on src/pages/landing/components/RegisterSection.tsx:
```diff
@@ -0,0 +1,168 @@
+            <TextField
+              className="mb-6"
+              required fullWidth variant="standard"
+              label="Email" type="email"
```
```suggestion
              label="Company Email" type="email"
```

RSoulatIOHK commented on src/pages/landing/components/RegisterSection.tsx:
```diff
@@ -0,0 +1,168 @@
+            <TextField
+              className="mb-6"
+              required fullWidth variant="standard"
+              label="Contact email" type="email"
```
```suggestion
              label="Contact email" type="email"
```

RSoulatIOHK commented on src/pages/landing/components/RegisterSection.tsx:
```diff
@@ -0,0 +1,146 @@
+              label="Twitter" type="text"
+              error={errors.twitter !== undefined}
+              helperText={errors.twitter !== undefined ? 'The field value is invalid' : undefined}
+              {...register("twitter", { pattern: /^[A-Za-z0-9_]{1,15}$/i })}
```
I think this is a good suggestion. Thanks @coderabbitai

amnambiar commented:
Approving from my side, @EehMauro need to address the suggestions by Romain though

EehMauro committed:
Added documentation link (#18)

Co-authored-by: Mauro Escudero <eehmauro@EehMacBookProM1Max.local>

EehMauro committed:
Update src/pages/landing/components/RegisterSection.tsx

Co-authored-by: RSoulatIOHK <117812549+RSoulatIOHK@users.noreply.github.com>

Commit:
Fixed registration form price

RSoulatIOHK commented:
Looks good! Thanks
